DRAFT_USAGE.md

```diff
@@ -151,15 +185,70 @@ proc handler(request: Request) {.async.} =
 ```
 
-## Authentication HttpCode401
+## Basic Authentication
+
+Incomplete API. This is subject to change. See src/scorper/http/httpbasicauth
+
+A request can have authorization basic headers automatically decoded and then passed to a validator for verification.
+
 ```nim
+# Create validator
+# Must take a Request and two string params and return a Future[bool]
+
+proc basicValidation(request: Request, user, pass: string): Future[bool] {.async.} =
+  # Check the decoded user and pass against your database/whatever
+  # return true for authentication to be accepted and for the calling callback to continue
+  # return false for automatic Http401 response
+  return true
+
+# Create your callback
 proc handler(request: Request) {.async.} =
-  if req.headers.hasKey("Authorization"):
-    await req.resp("")              # Client will receive a Http200
+  # Create validation object and pass to request method basicAuth
+  var validator = HttpBasicAuthValidator(basicValidation) # our validator proc
+  if await request.basicAuth(validator):
+    # Do things because request is validified
   else:
-    await req.respBasicAuth()       # Client will receive a Http401
+    # Do things here because request was invalid
+    # RESPONSE UNNECESSARY; see note below
+
+  # basicAuth proc returns a bool; however any failure to validate the auth header
+  # will result in an automatic response of either Http400 or Http401. An if/else
+  # statement can be used however responses for failure are unnecessary.
+  # Using the statement below is therefore perfectly valid
+  #
+  #   if not await request.basicAuth(validator): return
+
+# Can create router here or set server handler directly
+let address, flags = "127.0.0.1:8888", {ReuseAddr}
+let server = newScorper(address, handler, flags)
+server.start()
+waitFor server.join()
+```
+
+## Template responses
+
+A variety of procs can be used to generate responses to requests. The procedures and their parameters are listed below; please see src/scorper/http/streamserver for their implementations.
+
+Listed below are some common response procs.
+
+```nim
+proc respBasicAuth*(req: ImpRequest, scheme = "Basic", realm = "Scorper", params: seq[tuple[key: string,
+    value: string]] = @[], code = Http401): Future[void] {.async.}
+    ## Responds to the req with the specified ``HttpCode``; defaults to Http401
+
+proc respError*(req: ImpRequest, code: HttpCode, content: sink string, headers = newHttpHeaders()): Future[
+    void] {.async.}
+proc respError*(req: ImpRequest, code: HttpCode, headers = newHttpHeaders()): Future[void] {.async.}
+  ## Responds to the req with the specified ``HttpCode``.
+
+proc respStatus*(req: ImpRequest, code: HttpCode, ver = HttpVer11): Future[void] {.async.}
+proc respStatus*(req: ImpRequest, code: HttpCode, msg: string, ver = HttpVer11): Future[void] {.async.}
+
+## Chore: include responses related to files/downloads etc
 ```
 
+<!--- Note: request default respError codes --->

```

Comment:
RespError would be expected to have a default HttpCode if none is passed; perhaps defaulting to a Http400.

What do you think?

Reply:
I see; you are right, I remember that coming up in my testing. I suppose I am more indicating if within my handler I wanted to return an error for whatever reason, using respError would be my initial thought as to the appropriate procedure to use. I would personally overload this anyway to have a default code while in development.

- [ ] Add documentation about default error handler magic

Reply:
I haven't need this feature from my experience. It seems all server and frameworks response 500 due to internal program error. well If you have better design of default error code for helping end user development, welcome to create a PR .